projects/graph/src/graph.py

```diff
@@ -1,9 +1,269 @@
+#     # create an empty queue
+#     q = Queue()
+#     # Put starting vert in the queue
+#     q.enqueue(starting_node)
+#     while q.size() > 0:  # whlie queue is not empty...
+#         dequeued = q.dequeue() # Dequeue the first element
+#         visited.append(dequeued)  # Mark it as visited
+#         print(dequeued)
+#         if dequeued == target_node:
+#             return True
+#         for edge in self.vertices[dequeued].edges:  #For each child
+#             if edge not in visited:  # If it hasn't been visited
+#                 q.enqueue(edge)  # Add it to the back of the queue
+#     return False
+
+# def dfs(self, starting_node, target_node, visited=None):
+#     """
+#     Depth first traversal using recursion
+#     """
+#     # Mark the node as visited
+#     if visited is None:
+#         # quese of visited nodes
+#         visited = []
+#     visited.append(starting_node)
+#     if starting_node == target_node:
+#         return True
+#     # For each child, if that child hasn't been visited, call dft() on that node
+#     for node in self.vertices[starting_node].edges:
+#         if node not in visited:
+#             if self.dfs(node, target_node, visited):
+#                 return True
+#     return False
+
+
+# class Vertex:
+#     def __init__(self, vertex_id, x=None, y=None):
+#         """
+#         Create an empty vertex
+#         """
+#         self.id = vertex_id
+#         self.edges = set()
+#         if x is None:
+#             self.x = random.random() * 10 - 5
+#         else:
+#             self.x = x
+#         if y is None:
+#             self.y = random.random() * 10 - 5
+#         else:
+#             self.y = y
+
+#     def __repr__(self):
+#         return f"{self.edges}"
+
+##############################################################
+##############################################################
 class Graph:
-    """Represent a graph as a dictionary of vertices mapping labels to edges."""
     def __init__(self):
-        pass  # TODO
+        pass  # TODO	        self.vertices = {}
+        self.components = 0
+
+    def add_vertex(self, vertex, edges=()):
+        """Add a new vertex, optionally with edges to other vertices."""
+        if vertex in self.vertices:
+            raise Exception("Error: adding vertex that already exists")
+        if not set(edges).issubset(self.vertices):
+            raise Exception("Error: cannot have edge to nonexistent vertices")
+        self.vertices[vertex] = set(edges)
+
+    def add_edge(self, start, end, bidirectional=True):
+        """Add a edge (default bidirectional) between two vertices."""
+        if start not in self.vertices or end not in self.vertices:
+            raise Exception("Vertices to connect not in graph!")
+        self.vertices[start].add(end)
+        if bidirectional:
+            self.vertices[end].add(start)
+
+    def search(self, start, target=None, method="dfs"):
+        """Search the graph using BFS or DFS."""
```

Comment:
If you take nothing else away from this week, remember what you did in this file. BFS and DFS are some of the most common interview questions and being able to show how they work and explain it can mean the difference in many situations, whether it's a job or a salary negotiation, knowing this makes you that much more valuable.